### PR TITLE
Review: thedracle/node-webrtc#mediastream-rebase-squash

### PR DESCRIPTION
**Pull request by markandrus**
@thedracle I'm opening this PR so that it will be easier to review which of your changes we can land. I think it may make sense to open a few separate PRs with subsets of these changes.

### CODE REVIEW

File: lib/mediastream.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+RTCMediaStream.prototype.RTCMediaStates = [
+  'connecting',
+  'open',
+  'closing',
+  'closed'
+];
+
+module.exports = RTCMediaStream;
```
Comment (markandrus) [inline lib/mediastream.js:66]: This looks roughly like what we will need for MediaStream. In #393, I've added MediaStreamTrack and tried to avoid the JavaScript wrapper (everything's implemented in the C++) layer; however, I haven't added—nor do I have a good pattern for—EventTarget support (which we need for both MediaStream and MediaStreamTrack). Until we can handle EventTarget support in C++, I think we can keep the JavaScript wrappers.

File: lib/mediadevices.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  this.getUserMedia = function getUserMedia(successCallback, errorCallback){
+    var stream = md.getUserMedia();
+    if (stream === null) {
+      return errorCallback(new Error("Couldn't create MediaStream"));
+    }
+    successCallback(new RTCMediaStream(stream));
+  };
+}
+
+module.exports = MediaDevices;
```
Comment (markandrus) [inline lib/mediadevices.js:75]: I think it should be possible to remove the JavaScript wrapper here? Also, I don't think we will need the `executeNext`/`queueOrRun` infrastructure going forward. It's been removed from `lib/peerconnection.js`, for example.

File: lib/mediastreamevent.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      enumerable: true
+    },
+    'stream': {
+      value: eventInitDict.stream,
+      enumerable: true
+    }
+  });
+}
+
+module.exports = RTCMediaStreamEvent;
```
Comment (markandrus) [inline lib/mediastreamevent.js:15]: We may just want to use RTCTrackEvent (implemented in #393), since `onaddstream` is not present in the recent WebRTC draft spec. We could then consider shimming legacy APIs in the JavaScript layer.

File: lib/peerconnection.js [hunk 4]
```diff
@@ -81,39 +85,145 @@ function RTCPeerConnection(configuration, constraints) {
...
+    self.dispatchEvent(videoFrameEvent);
+    self._onvideoframe && self._onvideoframe(videoFrameEvent);
+  };
+
+  pc.onencodedvideoframe = function(id, width, height, encodedFrame, frameType, timestamp) {
+      var videoFrameEvent = new RTCVideoFrameEvent('videoframe',
+          {id: id, width: width, height: height, encodedFrame: encodedFrame,
+              frameType: frameType, timestamp: timestamp});
+      self.dispatchEvent(videoFrameEvent);
+      self._onencodedvideoframe && self._onencodedvideoframe(videoFrameEvent);
```
Comment (markandrus) [inline lib/peerconnection.js:159]: I think an issue with the `onaudioframe`/`onvidoeframe`/`onencodedvideoframe` will be that it doesn't make it easy to distinguish frames for different tracks of the same kind? Unless `id` will match the underlying MediaStreamTrack's `id`? I'm not sure how this works yet. **Edit:** I see below now it does.

IMO, it would be better if these events were raised on the MediaStreamTracks themselves, rather than the RTCPeerConnection. IIUC, `onencodedvideoframe` should make sense on a local video MediaStreamTrack even if its not yet added to an RTCPeerConnection.

File: lib/peerconnection.js [hunk 5]
```diff
@@ -340,7 +450,38 @@ function RTCPeerConnection(configuration, constraints) {
...
+              typeof(arguments[1]) === 'function' &&
+              typeof(arguments[2]) === 'function') {
+              // Legacy method.
+              call(stream, arguments[1], arguments[2]);
+          }
+          else {
+              throw error;
+          }
+      }
+  }
```
Comment (markandrus) [inline lib/peerconnection.js:482]: Let us use `addTrack`, once it is supported, and shim `addStream` in terms of that (or delegating to libwebrtc, if it does the shimming itself). But, in either situation, this method should be `void` AFAICT.

File: lib/rtcaudioframeevent.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            enumerable: false
+        },
+        'buffer': {
+            value: eventInitDict.buffer,
+            enumerable: true
+        }
+    });
+}
+
+module.exports = RTCAudioFrameEvent;
```
Comment (markandrus) [inline lib/rtcaudioframeevent.js:35]: This looks useful 👍we may want to use camelCase for the attributes, though.

File: lib/rtcvideoframeevent.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            enumerable: true
+        },
+        'vPlane': {
+            value: eventInitDict.vPlane,
+            enumerable: true
+        }
+    });
+}
+
+module.exports = RTCVideoFrameEvent;
```
Comment (markandrus) [inline lib/rtcvideoframeevent.js:45]: This also looks useful 👍same comment as above.

File: src/decoderproxy.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+#define NODEVP8_DECODER_PROXY_H
+
+#include "webrtc/modules/video_coding/codecs/vp8/include/vp8.h"  // nogncheck
+#include "webrtc/modules/video_coding/utility/ivf_file_writer.h"
+#include <memory>
+#include <string>
+#include <functional>
+#include <map>
+
+class DecoderProxy : public webrtc::VideoDecoder {
```
Comment (markandrus) [inline src/decoderproxy.h:11]: Why is this necessary? I may have to read up...

File: src/mediadevices.cc [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+using v8::String;
+using v8::Uint32;
+using v8::Value;
+
+Nan::Persistent<Function> MediaDevices::constructor;
+
+
+// MediaDevices constructor/destructor
+MediaDevices::MediaDevices() {
+  _peer_connection_factory = PeerConnectionFactory::GetOrCreateDefault();
```
Comment (markandrus) [inline src/mediadevices.cc:30]: PeerConnectionFactory should allow choosing the AudioLayer it uses to `Create` AudioDeviceModule. Once we do this, it may make sense to pass along the PeerConnectionFactory (specifying a particular AudioLayer) to classes like PeerConnection and MediaDevices.

We probably also want a static `getUserMedia` that lazily constructs MediaDevices, not unlike PeerConnection, if possible? Not sure if that'd complicate the code, but it'd mean

```js
const { getUserMedia } = require('wrtc')

(async () => {
  const stream = getUserMedia({ audio: true, video: true })
  // PeerConnectionFactory constructed
  stream.getTracks().forEach(track => track.stop())
})().then(() => {
  // PeerConnectionFactory destroyed
})
```

just as the following works today:

```js
const { RTCPeerConnection } = require('wrtc')

(async () => {
  const pc = new RTCPeerConnection()
  // PeerConnectionFactory constructed
  pc.close()
})().then(() => {
  // PeerConnectionFactory destroyed
})
```

File: src/mediadevices.cc [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+NAN_METHOD(MediaDevices::GetUserMedia) {
+  TRACE_CALL;
+
+  MediaDevices* self = Nan::ObjectWrap::Unwrap<MediaDevices>(info.This());
+
+  const char kAudioLabel[] = "audio_label";
+//  const char kVideoLabel[] = "video_label";
+  const char kStreamLabel[] = "stream_label";
```
Comment (markandrus) [inline src/mediadevices.cc:63]: We'll want to generate UUIDs, I think.

File: src/mediadevices.cc [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      self->_peer_connection_factory->factory()->CreateLocalMediaStream(kStreamLabel);
+
+  stream->AddTrack(audio_track);
+//  stream->AddTrack(video_track);
+
+  Local<Value> cargv[1];
+  cargv[0] = Nan::New<External>(static_cast<void*>(stream));
+  Local<Value> obj = Nan::New(MediaStream::constructor)->NewInstance(1, cargv);
+
+  info.GetReturnValue().Set(obj);
```
Comment (markandrus) [inline src/mediadevices.cc:85]: We should construct a `v8::Promise::Resolver`, call `Resolve` with the MediaStream, and then return the result of calling `GetPromise` on the `v8::Promise::Resolver`.

File: src/mediadevices.cc [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  TRACE_END;
+}
+
+void MediaDevices::Init(Handle<Object> exports) {
+  Local<FunctionTemplate> tpl = Nan::New<FunctionTemplate>(New);
+  tpl->SetClassName(Nan::New("MediaDevices").ToLocalChecked());
+  tpl->InstanceTemplate()->SetInternalFieldCount(1);
+
+  Nan::SetPrototypeMethod(tpl, "getUserMedia", GetUserMedia);
```
Comment (markandrus) [inline src/mediadevices.cc:95]: Seems like we'll want this (so that you could have two MediaDevices using different AudioLayers), although we may want a static `getUserMedia` that uses the default PeerConnectionFactory (see comment above).

File: src/mediastream.cc [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+using v8::Uint32;
+using v8::Value;
+
+Nan::Persistent<Function> MediaStream::constructor;
+
+MediaStream::MediaStream(webrtc::MediaStreamInterface* msi)
+  : loop(uv_default_loop()),
+    _internalMediaStream(msi) {
+  uv_mutex_init(&lock);
+  uv_async_init(loop, &async, reinterpret_cast<uv_async_cb>(Run));
```
Comment (markandrus) [inline src/mediastream.cc:33]: We'll want to convert to EventLoop rather than managing libuv directly.

File: src/mediastream.cc [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+    }
+    if(MediaStream::ADDTRACK & evt.type) 
+    {
+
+    }
+    if(MediaStream::REMOVETRACK & evt.type) 
+    {
+
+    }
```
Comment (markandrus) [inline src/mediastream.cc:119]: One of my questions with adding MediaStream (related to #396) is this: when a MediaStreamTrackInterface is added to a MediaStreamInterface by WebRTC, how do we figure out if it needs to be wrapped in our own MediaStreamTrack wrapper or not? We don't just want to re-wrap in a new MediaStreamTrack every time. Maybe the MediaStreamTrack class should have a static map from MediaStreamTrackInterfaces to MediaStreamTracks, and we just do a "get or create" operation for a new MediaStreamTrack? Then I'm just worried about leaks, but probably these can be removed in response to some event.

File: src/mediastream.cc [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+                                   == track->GetSource()->kMuted));
+
+    // ToDO: Determine if track is read only.
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("readonly").ToLocalChecked(),
+             Nan::New<v8::Boolean>(true));
+
+    Nan::Set(jsAudioTrack, Nan::New<v8::String>("remote").ToLocalChecked(),
+             Nan::New<v8::Boolean>(track->GetSource()->remote()));
+
+    Nan::Set(audioTracksJSArray, Nan::New<Number>(i++), jsAudioTrack);
```
Comment (markandrus) [inline src/mediastream.cc:165]: We should be able to update this to use the new `CONVERT_OR_THROW_AND_RETURN` macro.

File: src/mediastream.cc [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+                                   == track->GetSource()->kMuted));
+
+    // ToDO: Determine if track is read only.
+    Nan::Set(jsVideoTrack, Nan::New<v8::String>("readonly").ToLocalChecked(),
+             Nan::New<v8::Boolean>(true));
+
+    Nan::Set(jsVideoTrack, Nan::New<v8::String>("remote").ToLocalChecked(),
+             Nan::New<v8::Boolean>(track->GetSource()->remote()));
+
+    Nan::Set(videoTracksJSArray, Nan::New<Number>(i++), jsVideoTrack);
```
Comment (markandrus) [inline src/mediastream.cc:202]: Same comment as above.

File: src/mediastreamtrack.cc [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+NAN_METHOD(MediaStreamTrack::stop) {
+  TRACE_CALL;
+
+  // todo: implement
+
+  TRACE_END;
+  info.GetReturnValue().Set(Nan::Undefined());
+}
```
Comment (markandrus) [inline src/mediastreamtrack.cc:103]: The implementation in #393 lacks support for this. We should add it.

File: src/mediastreamtrack.cc [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+
+NAN_GETTER(MediaStreamTrack::GetLabel) {
+  TRACE_CALL;
+
+  MediaStreamTrack* self = Nan::ObjectWrap::Unwrap<MediaStreamTrack>( info.Holder() );
+  std::string label = self->_internalMediaStreamTrack->id();
+
+  TRACE_END;
+  info.GetReturnValue().Set(Nan::New(label.c_str()).ToLocalChecked());
+}
```
Comment (markandrus) [inline src/mediastreamtrack.cc:123]: The implementation in #393 lacks support for this. We should add it.

File: src/mediastreamtrack.cc [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+
+NAN_METHOD(MediaStreamTrack::clone) {
+  TRACE_CALL;
+
+  // todo: implement
+
+  TRACE_END;
+  info.GetReturnValue().Set(Nan::Undefined());
+}
```
Comment (markandrus) [inline src/mediastreamtrack.cc:94]: The implementation in #393 lacks support for this. We should add it.

File: src/mediastreamtrack.cc [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+
+NAN_GETTER(MediaStreamTrack::GetMuted) {
+  TRACE_CALL;
+
+  // todo: implement
+
+  TRACE_END;
+  info.GetReturnValue().Set(Nan::New(false));
+}
```
Comment (markandrus) [inline src/mediastreamtrack.cc:153]: The implementation in #393 lacks support for this. We should add it.

File: src/peerconnection.cc [hunk 21]
```diff
@@ -181,6 +192,155 @@ void PeerConnection::Run(uv_async_t* handle, int status) {
...
+            }
+
+            self->QueueEvent(PeerConnection::NOTIFY_ON_AUDIO_FRAME, audioFrameEvent);
+
+            audioSampleBuffer->assign(audioSampleBuffer->begin() + AUDIO_SAMPLE_DELIVERY_SIZE, audioSampleBuffer->end());
+          }
+        }, label);
+        audioTrack->GetSource()->AddSink(sink);
+      }
+    }
```
Comment (markandrus) [inline src/peerconnection.cc:339]: We should convert these to `Handle*` methods (newer style introduced in #387).

File: src/peerconnection.cc [hunk 22]
```diff
@@ -193,6 +353,99 @@ void PeerConnection::Run(uv_async_t* handle, int status) {
...
+  auto videoTracks = mediaStreamInterface->GetVideoTracks();
+  string label = mediaStreamInterface->label();
+  for(auto videoTrack : videoTracks) {
+    videoTrack->AddOrUpdateSink(new rtc::RefCountedObject<VideoSink>([label, self](const webrtc::VideoFrame& frame, string inLabel) {
+      VideoFrameEvent* event = new VideoFrameEvent(label);
+      event->width = frame.width();
+      event->height = frame.height();
+      event->buffer = frame.video_frame_buffer();
+      self->QueueEvent(PeerConnection::NOTIFY_ON_FRAME, event);
+    }, label), rtc::VideoSinkWants());
```
Comment (markandrus) [inline src/peerconnection.cc:375]: How should this sink be removed?

File: src/peerconnection.cc [hunk 23]
```diff
@@ -193,6 +353,99 @@ void PeerConnection::Run(uv_async_t* handle, int status) {
...
+
+      event->timestamp = bigPts;
+      wrapAroundData.lastTimeStamp = event->timestamp;
+
+      event->width = frame._encodedWidth;
+      event->height = frame._encodedHeight;
+
+      event->buffer.insert(event->buffer.begin(), frame._buffer, frame._buffer + frame._length);
+      self->QueueEvent(PeerConnection::NOTIFY_ON_ENCODED_FRAME, event);
+    });
```
Comment (markandrus) [inline src/peerconnection.cc:442]: Similar question as above: should this be/is this unregistered?

File: lib/mediastream.js (hunk 1 above)
Comment (thedracle) [reply lib/mediastream.js:66]: Nice. I have a sort of mockup of MediaStreamTrack in my branch that just composes an JS object from, but it definitely seems better to avoid the wrapper.

Just a question--- does this make it so in javascript land a track would pass type checks against RTCMediaStreamTrack?

There is some nicety in being able to wrap and modify the interface to these things in Javascript rather than in C++, particularly if in the future we wanted to more easily provide support for spec changes without having to rewrite a significant portion of the C++ interface.

Comment (markandrus) [reply lib/mediastream.js:66]: > Just a question--- does this make it so in javascript land a track would pass type checks against RTCMediaStreamTrack?

Sorry, I'm not sure what you mean. Can you ask in a different way? With either approach, given some `mediaStreamTrack` and `mediaStream`, we can satisfy

```js
const { MediaStream, MediaStreamTrack } = require('wrtc')
mediaStream instanceof MediaStream
mediaStreamTrack instanceof MediaStreamTrack
```

if we want to. We can also ensure methods that should only accept proper instances of these accept only proper instances, too. There's just more wrapping and unwrapping necessary if we have JavaScript wrappers.

> There is some nicety in being able to wrap and modify the interface to these things in Javascript rather than in C++, particularly if in the future we wanted to more easily provide support for spec changes without having to rewrite a significant portion of the C++ interface.

Yes, this is true. For example, it might be nice to shim the legacy APIs in JavaScript rather than bind to the native APIs (I'm thinking `addStream`, or the callback versions of `setLocalDescription`, etc.—unless the shims are complicated enough and libwebrtc actually provides correct legacy APIs). As far as newly specced APIs, I'm not sure what's easier in JavaScript right now. But we _can_ still shim methods C++ classes' `prototype` if necessary.

**Comment by markandrus** [comment]
Although different, I think #472 enables similar use cases as this PR, at least for video. We can track audio work in #478. Although I didn't land changes from this PR, it was useful to see the implementation.
